### ISSUE TITLE
Microsoft.Data.Sqlite: Pool Connections

### ISSUE THREAD

**Issue by bricelam** [dotnet/efcore#13837]
Pooling connections will improve performance--especially when using encryption which makes opening a connection more expensive. (Part of https://github.com/aspnet/Microsoft.Data.Sqlite/issues/412)

This would also allow us to make `CreateFunction()`, etc. calls persist between closing and re-opening a connection obviating the need for https://github.com/aspnet/Microsoft.Data.Sqlite/issues/354.

**Comment by bricelam** [comment]
There are some PRAGMAs that change the functional behavior of the connection. If we add a pool, we'll need to consider how we handle these.

- case_sensitive_like
- foreign_keys
- ignore_check_constraints
- read_uncommitted
- recursive_triggers
- reverse_unordered_selects

There are other statements that alter the connection state, but they're primarily for performance tuning and don't affect functionality.

**Comment by bricelam** [comment]
Also, extensions (e.g. SpatiaLite) are enabled per-connection. I don't think they can be disabled aside from closing and re-opening the connection.

**Comment by bricelam** [comment]
I decided this probably isn't a breaking change since we can make the behavioral breaking changes as part of #13826, and add pooling later to improve performance.

**Comment by roji** [comment]
@bricelam,

> There are some PRAGMAs that change the functional behavior of the connection. If we add a pool, we'll need to consider how we handle these.

I have a similar situation with PostgreSQL, where various connection state can be changed with parameters. PostgreSQL does contain a `RESET ALL` mechanism mechanism which can be used to reset any changes, when returning it to the pool - I wonder if SQLite has anything similar.

(note that Npgsql has a `No Reset On Close` for extreme-perf-mode which does not do this)

**Comment by bricelam** [comment]
SQLite 3.30.0 added support to unload extensions: sqlite3_drop_modules

**Comment by DkozhevnikovConfigit** [comment]
Hi EFCore team, any news about connection pooling? Would it be implemented?

**Comment by ajcvickers** [comment]
@DkozhevnikovConfigit Work is in-progress.

**Comment by rrgurski** [comment]
Here is a really simple way of doing that, using a non-blocking approach.

Make sure to call ``SqliteConnectionPool.Dispose()`` when your app is terminating.

Currently I'm using .NET Core 3.1 with this code. Probably will work in newer versions as well.

SqliteConnectionPool.cs
```
using Microsoft.Data.Sqlite;
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.Common;

namespace YourNamespace
{
    public static class SqliteConnectionPool
    {
        private const int PoolSize = 10;

        private static readonly object Locker = new object();

        static SqliteConnectionPool()
        {
            Pool = new List<DbConnection>(PoolSize);
            Additional = new List<DbConnection>();
        }

        private static bool Finalized { get; set; }

        private static int CreatedConnectionsMax => PoolSize;
        private static int CreatedConnectionsCount { get; set; }

        private static List<DbConnection> Pool { get; }
        private static List<DbConnection> Additional { get; }

        public static DbConnection GetConnection()
        {
            if (Finalized)
            {
                throw new Exception("Connection pool was finalized.");
            }

            lock (Locker)
            {
                if (Pool.Count == 0)
                {
                    var connection = CreateAndOpenConnection();
                    if (CreatedConnectionsCount == CreatedConnectionsMax)
                    {
                        Additional.Add(connection);
                        LogAdditionalConnection();
                        return connection;
                    }
                    else
                    {
                        CreatedConnectionsCount++;
                        return connection;
                    }
                }
                else
                {
                    var connection = Pool[0];
                    Pool.RemoveAt(0);
                    return connection;
                }
            }
        }

        public static void ReleaseConnection(DbConnection connection)
        {
            if (Finalized)
            {
                CloseAndDisposeConnection(connection);
            }

            lock (Locker)
            {
                if (Additional.Contains(connection))
                {
                    Additional.Remove(connection);
                    CloseAndDisposeConnection(connection);
                }
                else
                {
                    Pool.Add(connection);
                }
            }
        }

        public static void Dispose()
        {
            lock (Locker)
            {
                foreach (var connection in Pool)
                {
                    CloseAndDisposeConnection(connection);
                }

                Finalized = true;
            }
        }

        private static DbConnection CreateAndOpenConnection()
        {
            // create connection string
            var connectionString = new SqliteConnectionStringBuilder()
            {
                DataSource = "YourPathToDatabaseFile",
                Password = "YourPassword",
                Mode = SqliteOpenMode.ReadWriteCreate,
            };

            // create and open connection
            var conn = new SqliteConnection(connectionString.ToString());
            conn.Open();

            // enable write-ahead log
            var cmd = conn.CreateCommand();
            cmd.CommandText = "PRAGMA journal_mode = 'wal'";
            cmd.ExecuteNonQuery();

            return conn;
        }

        private static void CloseAndDisposeConnection(DbConnection connection)
        {
            if (connection.State != ConnectionState.Closed)
            {
                connection.Close();
            }
            connection.Dispose();
        }

        private static void LogAdditionalConnection()
        {
            // it's important to know if your code is frequently opening additional connections
            // that way you can increase the pool size to improve performance
        }
    }
}
```

YourDbContext.cs
```
using Microsoft.EntityFrameworkCore;

namespace YourNamespace
{
    public class YourDbContext : DbContext
    {
        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            var connection = SqliteConnectionPool.GetConnection();

            optionsBuilder
                .UseLazyLoadingProxies()
                .UseSqlite(connection);
        }

        public override void Dispose()
        {
            var connection = Database.GetDbConnection();

            base.Dispose();

            SqliteConnectionPool.ReleaseConnection(connection);
        }
    }
}
```

**Comment by MacKey-255** [comment]
@rrgurski In any way can I control that the database is not locked if it is in use by other applications, using SqliteConnectionPool?
Isn't it more performace to have a single connection?

**Comment by rrgurski** [comment]
@MacKey-255 what do you mean by different applications? Different threads in the same program or different processes?

If it's other running programs I don't know if there is an easy way to do that and personally I would use a client/server database instead of a file database to do that.

Now if you are talking about the same program, with different threads, each thread must have at least a different connection for itself. You cannot share the same connection between threads, because they are not thread safe. It may apparently works, but you can get all kinds of strange behaviors and errors.

About the exception you mentioned, is definitely because you are trying to begin more than one transaction in the same connection:
```
var conn = SqliteConnectionPool.GetConnection();

var t1 = conn.BeginTransaction();
var t2 = conn.BeginTransaction(); // System.InvalidOperationException: SqliteConnection does not support nested transactions.
```

As a side note, [SQLite really does not support multiple write transactions at the same time](https://sqlite.org/lang_transaction.html#read_transactions_versus_write_transactions). But the connection provider internally implement some work around for this.

For example:
```
var conn1 = SqliteConnectionPool.GetConnection();
var conn2 = SqliteConnectionPool.GetConnection();

var t1 = conn1.BeginTransaction();
var t2 = conn2.BeginTransaction(); // this get blocked for several seconds until it throws: Microsoft.Data.Sqlite.SqliteException: 'SQLite Error 5: 'database is locked'.'
```

And this works:
```
var conn1 = SqliteConnectionPool.GetConnection();
var conn2 = SqliteConnectionPool.GetConnection();

var t1 = conn1.BeginTransaction();

_ = Task.Run(() =>
{
    var t2 = conn2.BeginTransaction(); // this get blocked until the t1 is finished

    var cmd2 = conn2.CreateCommand();
    cmd2.CommandText = "INSERT INTO ...";
    cmd2.ExecuteNonQuery();

    t2.Commit();
    t2.Dispose();
});

var cmd1 = conn1.CreateCommand();
cmd1.CommandText = "INSERT INTO ...";
cmd1.ExecuteNonQuery();

t1.Commit();
t1.Dispose();
```

In the most common scenarios, you will never get problems writing concurrently. Only if you are trying to do a lot of writes at the same time. In that case, develop a single thread to make the writes with a queue or just do not use SQLite.

**Comment by rrgurski** [comment]
@MacKey-255 reading again my response, I don't know if it was clear the response for your question "how to fix": you are probably using the same connection between more than one thread or you are not explicit finishing the transaction after begin the other one.

This works in the same connection:
```
var conn = SqliteConnectionPool.GetConnection();

using (var t1 = conn.BeginTransaction())
{
    // do stuff
    t1.Commit();
}

using (var t2 = conn.BeginTransaction())
{
    // do stuff
    t2.Commit();
}
```

Also, [make sure to enable write-ahead log on your connection](https://docs.microsoft.com/en-us/dotnet/standard/data/sqlite/async).

Sorry for posting all this here. I know it is not really a forum, but it was one of the links that appeared when I was trying to solve performance issues with EF + SQLite. Hope this helps some folks passing by.

**Comment by MacKey-255** [comment]
@rrgurski I'm trying to use multiple database providers and one of these is SQLite. Try to keep a single connection due to performance issues and locked database error. I work on a web (not is ASP.NET) and create a DbContext for each request but when sharing the same connection the mentioned error occurs. When I save information is when the only thing happens and I really don't know what I can do to solve it. SAlu2s

**Comment by rrgurski** [comment]
@MacKey-255 it is difficult to help without further information on how are you handling the requests. Giving the fact that it is indeed web, it is highly likely to use more than one thread to handle each request. I strongly advice you to use the connection pool sample that I gave and test if the problem still happen.

**Comment by MacKey-255** [comment]
@rrgurski I already used the connection pool and there were many requests that never responded :(
Apparently this was stuck too long and never ended. I am going to carry out a separate Test of the application emulating multi-threads to be able to check the errors in more detail because no error occurred apparently but the group of connections was completely blocked. I will go back to my old code with a single connection and try to define `BeginTransaction` when writing information to the database. SAlu2s

**Comment by rrgurski** [comment]
@MacKey-255 assuming that there is nothing wrong with your code, I'm imagining two possible problems.

First one: how much time your requests can take to complete? If they can take some time to run, make sure to execute the transaction only in the relevant peace of code that need it. If you begin the transaction at the beginning of each request, it will block all the other transactions to happen and they will timeout with an exception in a few seconds (like the example I posted above).

Second one: how many requests can your application process in a short period of time? If there is too many requests, the transactions will be blocking each other all the time and some of them can timeout before getting the chance to run.

Also, have you increased the pool size according to your needs? The pool sample I gave was originally targeting single user applications. The default size is only 10 connections at the same time, more than enough for a common desktop app for example.

If you reach this limit, the pool start creating additional connections each time, closing then when the operation is over. This is done because it's a simple implementation and it's easier than blocking the execution until one connection is returned.

In other words, if you do not adjust the pool size and you are using more than the 10 connections at the same time, your application will slow down the same way it did without the pool (maybe causing the requests to timeout?).

To adjust the pool size (available open connections at the same time):
```
public static class SqliteConnectionPool
{
    private const int PoolSize = 10; // modify here
    //...
}
```

Finally, I will reinforce it again: if you are using more than one thread to process the requests **DO NOT use a single shared connection**. You can read [here](https://github.com/dotnet/efcore/issues/22664#issuecomment-696870423) and [here](https://github.com/dotnet/efcore/issues/22664#issuecomment-697183716) about it in the other issue you commented.

**Comment by MacKey-255** [comment]
@rrgurski
1) Usually requests take 50ms-100ms or less.
I do not use transactions directly, they are used automatically when I save information with `SaveChanges()`. Even if I try to use a transaction and do a `SaveChanges()` it throws the error: `System.InvalidOperationException: SqliteConnection does not support nested transactions`. :(

2) The number of requests varies in terms of the number of users, let's say there is an average of 30-60 users constantly and each one makes 1 request every 1 second because they also access the WebSocket and I need to be constantly using data from the database.
The first thing I did was increase the pool of connections to 100 because I require them and it not throwing any visible errors. For that reason I assumed that the connection pool was totally blocked.

On the one hand I have the error of locked database when I use concurrency and on the other I have the problem with transactions when I share a single connection.

I need to maintain performace and not error: database is locked. I know I should use MySQL or some other database provider but I'm obliged to add SQLite too. So, What can I do?
SAlu2s

**Comment by rrgurski** [comment]
@MacKey-255 if all these requests write changes to the database, I would say you have a big problem. SQLite was not made for concurrent writes. But [somebody some time ago said that anything less than 100K page views a day could be handled perfectly in typical scenarios](https://stackoverflow.com/a/4060838/14166088).

One thing is for sure: with this amount of requests, you will definitely need a connection pool and one that is better than mine. Read the "**Update 2**" of my first comment with the code sample.

About what you can do, the only solution I can think of is to synchronize all the writes. If you think about it, it is already synchronized by the database itself, you just concurrently try to write and get blocked for doing that.

You can either put a lock in every place that writes to the database (if it's only ``SaveChanges()`` you can override it) or create a single thread with a queue to synchronously handle all the writes.

If your hardware is good enough, maybe it can work. But keep in mind that some day, if your customers/requests increase, it can stop working properly.

There is another work around: you can have multiple SQLite files if your data can be divided by customers or some other thing. It would only work if you do not have use cases that you need all the data together.

Anyway, I would rethink whether SQLite is the best choice here. In the future, if you have to change the database, it could be a nightmare.

**Comment by rrgurski** [comment]
@MacKey-255 I did some tests today regarding possible timeouts on write operations.

As far as I know, if you do not explicit create a transaction, EF creates one for each ``SaveChanges()`` you call. In order to simulate a timeout, I wrote this:

```
var conn1 = SqliteConnectionPool.GetConnection();
var conn2 = SqliteConnectionPool.GetConnection();

var t1 = conn1.BeginTransaction();

_ = Task.Run(() =>
{
    var t2 = conn2.BeginTransaction(); // this get blocked until the t1 is finished

    var cmd2 = conn2.CreateCommand();
    cmd2.CommandText = "INSERT INTO ...";
    cmd2.ExecuteNonQuery();

    t2.Commit();
    t2.Dispose();
});

await Task.Delay(29000);

var cmd1 = conn1.CreateCommand();
cmd1.CommandText = "INSERT INTO ...";
cmd1.ExecuteNonQuery();

t1.Commit();
t1.Dispose();
```

Look at this line of code ``await Task.Delay(29000)``. If I put 29 seconds it will always complete both transactions. But when I set it to 31 seconds, it always fail with a ``Microsoft.Data.Sqlite.SqliteException: 'SQLite Error 5: 'database is locked'.'``.

The conclusion is that every requested write operation can wait 30 seconds to be completed. In my case, I will never put a lock on my code, because if it is not completed in 30 seconds, it is a fail anyway.

For your case, assuming that is the same problem because of the exception you got, if you are stacking up requests in a way that they can take up more than 30 seconds to complete, either SQLite is a no go for you or there is something wrong with the code.

**Comment by MacKey-255** [comment]
@rrgurski After days of testing, this is my solution: 

SqliteConnectionPool.cs
```
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.Common;
using System.Linq;
using System.Threading;
using Microsoft.Data.Sqlite;

namespace YourNamespace
{
    /// <summary>
    /// SQLite Connection Pool Control.
    /// </summary>
    public class SqliteConnectionPool
    {
        #region Public-Members

        /// <summary>
        /// Logger intarnal activity (Used in testing)
        /// </summary>
        public Action<string> Logger;

        #endregion

        #region Private-Members

        /// <summary>
        /// Pool Size (1 = Single Connection)
        /// </summary>
        private const int PoolSize = 10;

        private readonly string _ConnectionString;
        private readonly SemaphoreSlim _ConnectionLock;

        /// <summary>
        /// State finalized Connection Pool
        /// </summary>
        private bool _Finalized { get; set; }
        private int _ConnectionCount { get; set; }
        private int _CreatedConnectionsMax => PoolSize;

        /// <summary>
        /// SQLite Connections and Transaction Locker
        /// </summary>
        private Dictionary<DbConnection, TransactionLocker> Pool { get; set; }

        #endregion

        #region Public-Methods

        /// <summary>
        /// Initializes the SQLite Connection Pool.
        /// </summary>
        /// <param name="connection">SQLite Connection string.</param>
        public SqliteConnectionPool(string connection)
        {
            _Finalized = false;
            _ConnectionCount = 0;
            _ConnectionString = connection;
            _ConnectionLock = new SemaphoreSlim(_CreatedConnectionsMax, _CreatedConnectionsMax);
            Pool = new Dictionary<DbConnection, TransactionLocker>(PoolSize);
        }

        /// <summary>
        /// Get new DbConnection from Connection Pool.
        /// </summary>
        public DbConnection GetConnection()
        {
            if (_Finalized)
            {
                throw new OperationCanceledException("Connection pool was finalized.");
            }

            // Single Connection
            if (_CreatedConnectionsMax == 1)
            {
                // Get Unique Open Connection
                if (Pool.Count > 0)
                    return Pool.First().Key;
                else
                {
                    // Create Unique Open Connection
                    var connection = CreateAndOpenConnection();
                    lock (Pool)
                    {
                        int id = _ConnectionCount++;
                        Pool.Add(connection, new TransactionLocker(id));
                        Logger?.Invoke("New single connection #" + id.ToString() + " added to Connection Pool");
                        return connection;
                    }
                }
            }
            else
            {
                _ConnectionLock.Wait(); // Wait until connection pool is released.

                // Double-Check Finalized
                if (_Finalized)
                {
                    throw new OperationCanceledException("Connection pool was finalized.");
                }

                // Add Connection to Pool
                var connection = CreateAndOpenConnection();
                lock (Pool)
                {
                    if (Pool.Count <= _CreatedConnectionsMax)
                    {
                        int id = _ConnectionCount++;
                        Pool.Add(connection, new TransactionLocker(id));
                        Logger?.Invoke("New single connection #" + id.ToString() + " added to Connection Pool");
                        return connection;
                    }
                    else
                    {
                        return connection;
                    }
                }
            }
        }

        /// <summary>
        /// Release DbConnection from Connection Pool.
        /// </summary>
        public void ReleaseConnection(DbConnection connection)
        {
            if (connection == null) return;
            if (_Finalized)
            {
                CloseAndDisposeConnection(connection);
                return;
            }

            // Single Connection (Not require release)
            if (_CreatedConnectionsMax == 1) return;

            try
            {
                if (_CreatedConnectionsMax > 1) // Double-check not single connection
                {
                    // Try Remove Connection
                    lock (Pool)
                    {
                        if (Pool.TryGetValue(connection, out var value))
                        {
                            // Remove Connection from Pool
                            Pool.Remove(connection);
                            Logger?.Invoke("Remove connection #" + value.ConnectionId.ToString() + " from Connection Pool");
                            CloseAndDisposeConnection(connection);
                        }
                    }
                }
            }
            catch (Exception ex) { Util.SaveException(ex, "Error Release Connection"); }
            finally
            {
                if (_ConnectionLock.CurrentCount < _CreatedConnectionsMax) _ConnectionLock.Release(); // Signal the connection pool to continue.
            }
        }

        /// <summary>
        /// Release All DbConnection from Connection Pool.
        /// </summary>
        public void ReleaseAllConnections()
        {
            // Close & Release All Connections
            lock (Pool)
            {
                foreach (var connection in Pool)
                {
                    CloseAndDisposeConnection(connection.Key);
                    Logger?.Invoke("Remove connection #" + connection.Value.ConnectionId.ToString() + " from Connection Pool");
                    if (_ConnectionLock.CurrentCount < _CreatedConnectionsMax) _ConnectionLock.Release(); // Signal the connection pool to continue.
                }
                Pool.Clear();
            }
        }

        /// <summary>
        /// Block Transaction in DbConnection.
        /// Soluction for: SqliteConnection does not support nested transactions.
        /// </summary>
        public void BlockTransaction(DbConnection connection)
        {
            if (connection == null) return;
            if (Pool.TryGetValue(connection, out var trans))
            {
                Logger?.Invoke("Blocked Transaction # " + trans.GetTransactionId().ToString() + " in Connection #" + trans.ConnectionId.ToString());
                trans.Locker.Wait(); // Wait until transaction in the same Connection is released.
            }
        }

        /// <summary>
        /// Release Transaction in DbConnection.
        /// Soluction for: SqliteConnection does not support nested transactions.
        /// </summary>
        public void ReleaseTransaction(DbConnection connection)
        {
            if (connection == null) return;
            if (Pool.TryGetValue(connection, out var trans))
            {
                Logger?.Invoke("Released Transaction # " + trans.GetTransactionId(true).ToString() + " in Connection #" + trans.ConnectionId.ToString());
                if (trans.Locker.CurrentCount < 1) trans.Locker.Release(); // Signal the transaction to continue.
            }
        }

        /// <summary>
        /// Tear down the Connection Pool and dispose of new Connections.
        /// </summary>
        public void Dispose()
        {
            _Finalized = true;
            ReleaseAllConnections();
            Pool = null;
            Logger?.Invoke("Disposed Connection Pool");
        }

        #endregion

        #region Private-Methods

        private DbConnection CreateAndOpenConnection()
        {
            // Create and Open Connection
            var conn = new SqliteConnection(_ConnectionString.ToString());
            conn.Open();

            // Enable write-ahead log
            var cmd = conn.CreateCommand();
            cmd.CommandText = "PRAGMA journal_mode = 'wal'";
            cmd.ExecuteNonQuery();

            return conn;
        }

        private void CloseAndDisposeConnection(DbConnection connection)
        {
            // Close and Dispose Connection
            if (connection == null) return;
            if (connection.State != ConnectionState.Closed)
            {
                connection.Close();
            }
            connection.Dispose();
        }

        /// <summary>
        /// Contain Connection ID and Semaphore Locker for the Transactions.
        /// </summary>
        private class TransactionLocker
        {
            public int ConnectionId { get; set; }
            private int TransactionId { get; set; }
            public SemaphoreSlim Locker { get; set; }

            public TransactionLocker(int id)
            {
                ConnectionId = id;
                TransactionId = 0;
                Locker = new SemaphoreSlim(1, 1);
            }

            public int GetTransactionId(bool plus = false)
            {
                int i = TransactionId;
                if(plus) TransactionId += 1;
                return i;
            }
        }

        #endregion
    }
}
```
This code allows me to manage the connections pool or a single connection. It avoid the error: `Microsoft.Data.Sqlite.SqliteException: 'SQLite Error 5: 'database is locked'.` and `System.InvalidOperationException: SqliteConnection does not support nested transactions`.
How to use: 

YourDbContext.cs
```
using Microsoft.EntityFrameworkCore;

namespace YourNamespace
{
    public class YourDbContext : DbContext
    {
        private readonly SqliteConnectionPool _Connection;

        public YourDbContext()
        {
            _Connection = DatabaseManager.ConnectionPool; // Get Static Class (Is important for share Class between all)
        }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            var connection = _Connection.GetConnection();

            optionsBuilder
                .UseLazyLoadingProxies()
                .UseSqlite(connection);
        }
        
        // It method use transactions automatically when save information
        public override int SaveChanges()
        {
            // Block Transiction for the same connectio to avoit error: SqliteConnection does not support nested transactions
            _Connection.BlockTransaction(GetConnection());
            int result = 0;
            try
            {
                result = base.SaveChanges();
            }
            catch { }
            finally
            {
                _Connection.ReleaseTransaction(GetConnection());
            }
            return result;
        }

        public override void Dispose()
        {
            var connection = Database.GetDbConnection();

            base.Dispose();

            SqliteConnectionPool.ReleaseConnection(connection);
        }
    }
}
```
SAlu2s
PS: Not all is tedted

**Comment by ajcvickers** [comment]
@sjb-sjb SqliteConnection inherits from DbConnetion. If you are using Sqlite, then your DbConnection is a SqliteConnection.

**Comment by ajcvickers** [comment]
@sjb-sjb How is this a breaking change?

**Comment by sjb-sjb** [comment]
The breaking change was that connection pooling results in the file being locked for longer than before. To unlock the file you need to call ClearPool. In my case when my UI program is in the background for a long time then I want to release the file so that OneDrive can copy it. The breaking change page talks about calling SqliteConnection.ClearPool but doesn't say how to get the SqliteConnection.

**Comment by sjb-sjb** [comment]
Fair enough. I guess the reason I was confused was I assumed that SqliteConnection would be provided by the sqlite engine rather than by EF Core, so it wasn't a natural conclusion for me that it would be derived from an EF class.

**Comment by ajcvickers** [comment]
DbConnection if not an EF Class.

**Comment by sjb-sjb** [comment]
I'm not trying to be hard-headed here. I know there is a breaking change note for EF, thanks, that's great. I'm just giving you the feedback that it would be nice if the breaking change note said "The SqliteConnection can be obtained by casting DbContext.Database.GetDbConnection() to SqliteConnection". It would save time poking around in the source code and bugging you nice folks in an attempt to figure it out. I know it's obvious to you but it wasn't very obvious to me for whatever reason.
